roachtest: add Datadog integration for test log ingestion

williamchoe3 opened the pull request:
Previously, we didn't have an easy way to do full text search across CI test runs across different TC Build Configurations and branches. To do that, you would have to download the artifacts for what you wanted to search for.

This change adds a datadog package that uploads test.log files to Datadog during test cleanup on master and release branches. The implementation scans the log file serially and uses a worker pool to upload log entries in batches of 1000 using the Datadog API client. Each log entry is tagged with test metadata (test name, owner, cloud, platform, version) and includes attributes for higher cardinality fields (cluster name, build number, result, duration). See comments for more details.

The entry point for datadog upload in `roachtest` will be during the post step test in `test_runner.go.`  Added a new roachtest flag `datadog-always-upload` for e2e testing on a non release branch.  Modified TC build scripts to pass new env vars and the teamcity build properties file.

golgeek commented on pkg/cmd/roachtest/datadog/datadog.go:
```diff
@@ -0,0 +1,461 @@
+	defer resp.Body.Close()
+	// Datadog returns 202 Accepted on success
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return errors.Newf("unexpected status code %d from Datadog API", resp.StatusCode)
```
Should we consider a retry mechanism for temporary errors?

golgeek reviewed (approved):
Looks great! Few comments here and there.

DarrylWong commented on pkg/cmd/roachtest/datadog/datadog.go:
```diff
@@ -0,0 +1,461 @@
+
+	// Start worker goroutines to upload batches concurrently
+	for i := 0; i < numWorkers; i++ {
+		g.GoCtx(func(ctx context.Context) error {
```
nit: you could pass in a `t.NewGroup` at the original call site in `runTest`. Would make this resilient to unrecovered panics, although my initial reading of the code it doesn't seem like a big concern.

williamchoe3 commented on pkg/cmd/roachtest/datadog/datadog.go:
```diff
@@ -0,0 +1,461 @@
+	// Start worker goroutines to upload batches concurrently
+	for i := 0; i < numWorkers; i++ {
+		g.GoCtx(func(ctx context.Context) error {
+			for batch := range batches {
```
Thx for the catch, added, will run a small smoke

williamchoe3 commented:
Smoke: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20892401?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

williamchoe3 committed:
roachtest: add Datadog integration for test log ingestion

Previously, test.log artifacts were only available on TeamCity's UI.

This change adds a datadog package that uploads test.log files to Datadog
during test cleanup on master and release branches. The implementation scans
the log file serially and uses a worker pool to upload log entries in batches
of 1000 using the Datadog API client. Each log entry is tagged with test
metadata (test name, owner, cloud, platform, version) and includes attributes
for higher cardinality fields (cluster name, build number, result, duration).

This enables full-text search across roachtest runs and improves observability
for test triage and failure analysis.

Epic: None
Release note: None

williamchoe3 commented:
Ripped out unnecessary "test.log" variable I initially added to minimize potential backport conflicts, would be nice to have a single place with all the consts used by roachtest, but can do that in a separate pr
